### DISCUSSION TITLE
faq/clb-loopback

### DISCUSSION THREAD

**Discussion by giscus** [imroc/tke-guide#6] [General]
https://imroc.cc/tke/faq/clb-loopback

**Comment by luolaifa000** [comment]
报文的源目的 IP 都在同一节点内，导致数据包在子机内部回环出
这个到达节点的时候  包被丢弃，源IP 是PODIP，然后目的节点IP。 这个是怎么看出来是本机IP，才丢弃的

PODIP 在node这一层 应该是看不到 pod的IP的，网卡好像只有cbr的IP

**Reply by luolaifa000** [reply]
包到节点的时候，怎么知道 PODIP 也是自己机器的IP的？

**Reply by imroc** [reply]
网桥模式的集群网络，每个节点有个podCidr容器网段，节点上会有路由，发给这个网段的包会转给cbr0，然后再由网桥二层转给节点上的pod

**Reply by luolaifa000** [reply]
通过clb，POD A 请求 到了 POD B，clb转发请求到节点上，然后这2个POD都在同一个节点上，节点上路由和网桥认为2个PODIP是 同一个容器网段的 就丢弃么

**Reply by luolaifa000** [reply]
那丢弃是节点的行为，还是PODA的行为

**Reply by imroc** [reply]
发包链路：POD A --> CLB --> POD B
回包链路：POD B --> POD A

来回都不在同一个链路，A只认识CLB，B直接回给A，会认为这个包不合法，就丢弃了

**Reply by luolaifa000** [reply]
还有一个疑问
报文到 NodePort 时，目的 IP 是节点 IP，源 IP 是 client pod 的真实 IP， 因为 CLB 不做 SNAT，会将真实源 IP 透传过去
假设podA 在节点1  podB在节点2 不是同一个节点哈
，如果clb不做snat，将包转发节点2，然后到达PODB，因为之前包的源地址是podA的，然后回包的发送IP 应该也是podA的，所以这里回包是直接回podA么，不会从CLB再原路返回么

**Reply by luolaifa000** [reply]
回包链路：POD B --> POD A 
这个回包如果直接是回给POD A的话，因为之前没建立过链接，POD A应该会RST，所以这里的RST就是上面文章说的丢弃行为么

**Reply by luolaifa000** [reply]
所以最后这个丢弃行为， 是POD A RST请求了，然后丢弃了么

**Reply by luolaifa000** [reply]
如果不再同一个节点上
回包链路：POD B --> POD A
节点2发现POD A 不是自己的容器网段，又会从CLB原路返回，这样就是正常回包了
理解了
非常感谢

**Reply by imroc** [reply]
> 还有一个疑问 报文到 NodePort 时，目的 IP 是节点 IP，源 IP 是 client pod 的真实 IP， 因为 CLB 不做 SNAT，会将真实源 IP 透传过去 假设podA 在节点1 podB在节点2 不是同一个节点哈 ，如果clb不做snat，将包转发节点2，然后到达PODB，因为之前包的源地址是podA的，然后回包的发送IP 应该也是podA的，所以这里回包是直接回podA么，不会从CLB再原路返回么

不是，CLB的底层实现用的隧道，会保证回包原路返回

**Reply by imroc** [reply]
> 所以最后这个丢弃行为， 是POD A RST请求了，然后丢弃了么

理论上应该是

**Comment by yongchuanzhou** [comment]
修改kube-proyx 的configmap 的 strictARP 参数是否能够绕过这个kube-ipvs网卡的arp应答？
